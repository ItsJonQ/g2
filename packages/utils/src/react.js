import { Children, isValidElement } from 'react';

/**
 * Merges React `ref` together.
 */
export { default as mergeRefs } from 'react-merge-refs';

/**
 * Copies non-react specific statics from a child component to a parent component.
 * Similar to Object.assign, but with React static keywords prevented from being overridden.
 */
export { default as hoistNonReactStatics } from 'hoist-non-react-statics';

/**
 * Gets a collection of available children elements from a React component's children prop.
 *
 * @param {import('react').ReactNode} children
 *
 * @return {import('react').ReactNodeArray} An array of available children.
 */
export function getValidChildren(children) {
	if (typeof children === 'string') return [children];

	return Children.toArray(children).filter((child) => isValidElement(child));
}

/**
 * Gets the displayName of a React Component or element.
 *
 * @param {string | import('react').ComponentType} tagName
 *
 * @return {string} The display name of the Component / tagName.
 */
export function getDisplayName(tagName) {
	const displayName =
		typeof tagName === 'string'
			? tagName
			: tagName.displayName || tagName.name || 'Component';

	return displayName;
}

/**
 * Checks to see if a `children` prop is a render function prop.
 *
 * @param {*} children
 *
 * @return {boolean} True, if children is a render function prop.
 */
export function isRenderProp(children) {
	return typeof children === 'function';
}

/**
 * Handles the rendering of a React component's children prop, which
 * may be a render function prop.
 *
 * @param {*} children Children to render.
 * @param {*} props Props to pass into a (potential) children render function.
 *
 * @return {*} The rendered children.
 */
export function renderChildren(children, props = {}) {
	if (isRenderProp(children)) {
		// eslint-disable-next-line no-unused-vars
		const { children: _, ...rest } = props;
		return children(rest);
	}

	return children;
}
