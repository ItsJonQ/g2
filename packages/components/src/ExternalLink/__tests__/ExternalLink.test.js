import { render } from '@testing-library/react';
import React from 'react';

import { ExternalLink } from '../index';

describe('props', () => {
	test('should render correctly', () => {
		const { container } = render(
			<ExternalLink href="https://wordpress.org">WordPress</ExternalLink>,
		);
		expect(container.firstChild).toMatchSnapshot();
	});
});
