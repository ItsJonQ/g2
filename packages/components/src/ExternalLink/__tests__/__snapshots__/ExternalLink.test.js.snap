// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`props should render correctly 1`] = `
.emotion-0.emotion-0.emotion-0.emotion-0.emotion-0.emotion-0.emotion-0 {
  box-sizing: border-box;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  font-family: Inter,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",sans-serif;
  font-family: var(--wp-g2-font-family);
  font-size: 13px;
  font-size: var(--wp-g2-font-size);
  font-weight: normal;
  font-weight: var(--wp-g2-font-weight);
  margin: 0;
  color: #1e1e1e;
  color: var(--wp-g2-color-text);
  line-height: 1.2;
  line-height: var(--wp-g2-font-line-height-base);
  font-size: calc((13 / 13) * 13px);
  font-size: calc((13 / 13) * var(--wp-g2-font-size));
  font-weight: normal;
  font-weight: var(--wp-g2-font-weight);
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #007cba;
  color: var(--wp-g2-link-color);
}

@media (prefers-reduced-motion) {
  .emotion-0.emotion-0.emotion-0.emotion-0.emotion-0.emotion-0.emotion-0 {
    -webkit-transition: none!important;
    transition: none!important;
  }
}

[data-system-ui-reduced-motion-mode="true"] .emotion-0.emotion-0.emotion-0.emotion-0.emotion-0.emotion-0.emotion-0 {
  -webkit-transition: none!important;
  transition: none!important;
}

.emotion-0.emotion-0.emotion-0.emotion-0.emotion-0.emotion-0.emotion-0:hover {
  color: #007cba;
  color: var(--wp-g2-link-color-hover);
}

.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1 {
  box-sizing: border-box;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  font-family: Inter,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",sans-serif;
  font-family: var(--wp-g2-font-family);
  font-size: 13px;
  font-size: var(--wp-g2-font-size);
  font-weight: normal;
  font-weight: var(--wp-g2-font-weight);
  margin: 0;
  display: block;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: currentColor;
  fill: currentColor;
  height: 15;
  width: 15;
  display: inline-block;
  vertical-align: middle;
  color: currentColor;
  margin-left: calc(4px * 1);
  margin-left: calc(var(--wp-g2-grid-base) * 1);
  margin-top: calc(4px * -1);
  margin-top: calc(var(--wp-g2-grid-base) * -1);
}

@media (prefers-reduced-motion) {
  .emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1 {
    -webkit-transition: none!important;
    transition: none!important;
  }
}

[data-system-ui-reduced-motion-mode="true"] .emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1 {
  -webkit-transition: none!important;
  transition: none!important;
}

.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1 svg {
  display: block;
}

<a
  class="components-truncate wp-components-truncate components-text wp-components-text ic-7u3orf ic-1m45iq3 components-link wp-components-link components-external-link wp-components-external-link ic-klt46f ic-euc9in ic-192yz5d emotion-0"
  data-g2-c16t="true"
  data-g2-component="ExternalLink"
  href="https://wordpress.org"
  target="_blank"
>
  <span
    style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
  >
    (opens in a new tab)
  </span>
  WordPress
  <div
    class="components-icon wp-components-icon ic-2q44rw ic-192yz5d emotion-1"
    data-g2-c16t="true"
    data-g2-component="Icon"
  >
    <svg
      aria-hidden="true"
      focusable="false"
      height="15"
      role="img"
      size="15"
      viewBox="0 0 24 24"
      width="15"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M18.2 17c0 .7-.6 1.2-1.2 1.2H7c-.7 0-1.2-.6-1.2-1.2V7c0-.7.6-1.2 1.2-1.2h3.2V4.2H7C5.5 4.2 4.2 5.5 4.2 7v10c0 1.5 1.2 2.8 2.8 2.8h10c1.5 0 2.8-1.2 2.8-2.8v-3.6h-1.5V17zM14.9 3v1.5h3.7l-6.4 6.4 1.1 1.1 6.4-6.4v3.7h1.5V3h-6.3z"
      />
    </svg>
  </div>
</a>
`;
